Improvement/use debug mode to rethrow exceptions

mrsasha opened the pull request:
This uses debug mode which was only partially implemented. It's now passed as a parameter to the `ViewBinder` constructor and is used by classes below. If `debugMode` is set to `true`, all exceptions will be logged and rethrown.

Implements https://github.com/alter-ego/androidbound/issues/157 and https://github.com/alter-ego/androidbound/issues/160.

mrsasha committed:
improved activity/fragment delegate logging in case of errors

mrsasha committed:
added null viewmodel check to BoundActivityDelegate.setContentView()

mrsasha committed:
removed IHasDebugMode, flag now passed in constructor

mrsasha committed:
surpressing RxLint onError messages in BindingAssociationEngine

mrsasha committed:
add debugMode to constructors in support module

mrsasha committed:
added debugMode exception rethrowing to ViewBinder & VBEngine

mrsasha committed:
added debugMode exception rethrowing to errors in BindingAssociationEngine

mrsasha committed:
added exception rethrowing to errors in BindableRecyclerViewAdapter

mrsasha committed:
added throwing exceptions on errors while binding

mrsasha committed:
adding debugMode exception rethrowing to warnings while binding

mrsasha committed:
adding debugMode exception rethrowing to BindableListAdapter

mrsasha committed:
adding debugMode exception rethrowing to ViewResolver

mrsasha committed:
adding debugMode exception rethrowing to BindingSpecificationParser

mrsasha committed:
enabled debug mode in example applications

mrsasha committed:
improved logging for some errors in binding and PropertyInfo

mrsasha committed:
fixed some method names in BindableRecyclerView

mrsasha committed:
improved ViewResolver logging

mrsasha committed:
fixed support examples for debug mode

mrsasha committed:
fixed example for debug mode

mrsasha committed:
fixed BindingSpecificationParser tests

mrsasha committed:
removed rethrowing when source is null in BindingAssociationEngine

when calling updateTargetFromSource(), if it would rethrow,
it would basically disable hierarchical binding if one of the
objects in the hierarchy is null. For that reason, we just print
it as an error instead of regular warning.

bblackbelt commented on AndroidBound/src/main/java/solutions/alterego/androidbound/android/viewresolvers/ViewResolver.java:
```diff
@@ -111,13 +117,17 @@ public View createView(String name, Context context, AttributeSet attrs) {
 
         try {
-            logger.debug("Resolving " + name + " with " + result);
+            logger.debug(getClass().getSimpleName() + " resolving " + name + " with " + result);
```
`getClass().getSimpleName()` is quite expensive. You could extract it a a constant

mrsasha replied:
I wanted it to print the name of the class, also those who extend it (like `SupportViewResolver`). Generally the idea is not to do logging in general when viewbinding unless debugging because it does slow things down.

bblackbelt replied:
you can still assign it to a constant.

bblackbelt commented on AndroidBound/src/main/java/solutions/alterego/androidbound/binding/BindingAssociationEngine.java:
```diff
@@ -125,12 +132,16 @@ public void accept(Exceptional<Object> value) throws Exception {
             } else {
-                mLogger.warning("Binding " + mBindingSpecification.getSource()
-                        + " needs Disposable, but changes were not available");
+                String msg = "Binding " + mBindingSpecification.getSource() + " needs Disposable, but changes were not available = " + mBindingSpecification.toString();
```
also, please avoid `+` for string concatenation and use `concat`

mrsasha replied:
I always thought that `+` gets converted to `StringBuilder` by the compiler, which is quicker when having bigger strings: https://stackoverflow.com/a/46485284

bblackbelt replied:
brain fart - sorry

bblackbelt commented on AndroidBound/src/main/java/solutions/alterego/androidbound/binding/ViewBindingEngine.java:
```diff
@@ -34,7 +34,7 @@
     private IImageLoader mImageLoader = IImageLoader.nullImageLoader;
 
-    private boolean mDebugMode;
+    private boolean mDebugMode = false;
```
it is false by default

mrsasha replied:
yes, just wanted to make it explicit.

bblackbelt commented on AndroidBound/src/main/java/solutions/alterego/androidbound/binding/types/CommandBinding.java:
```diff
@@ -76,7 +88,11 @@ private void retryExecute(View parameter, Object parameter2) {
                         mInfo.invoke(getSubject(), parameter2);
                     } catch (Exception e) {
-                        execute(parameter);
+                        if (mDebugMode) {
```
Why would you change this behaviour?

mrsasha replied:
Oh, you're right, good catch - it should go to a single parameter and try that one as the last resort. Removing it.

mrsasha committed:
removing throw when debugging in CommandBinding.retryExecute()

it will go to a single parameter execute()
and it will throw there, if it fails

mrsasha committed:
removed getClass().getSimpleName() from ViewResolver logging
